### PR TITLE
Introduce load balancing dataset samplers

### PR DESCRIPTION
**Pull request by baijumeswani**
This pull request introduces the `LoadBalancingDistributedSampler` which helps with straggler problems observed in distributed training tasks where different workers work on data samples with varying complexity resulting in stragglers.
The data sampler does the following:
- Sorts the dataset based on sample complexities.
- Distributes the data across workers.
- Shuffles the data each worker sees deterministically in order to avoid working with data that is sorted purely in ascending order (which may result in convergence issues).

Here is an example of what the data sampler does:
- Assume the data sampler is working with dataset with complexities: `[9, 8, 7, 6, 5, 4, 3, 2, 1, 0]` and the number of workers is 2.
- The sorted complexities will be: `[0 , 1, 2, 3, 4, 5, 6, 7, 8, 9]` and sorted dataset indices (sorted on complexities) will be `[9, 8, 7, 6, 5, 4, 3, 2, 1, 0]`.
- Data is then distributed among workers. So the two workers will work with the indices `[9, 7, 5, 3, 1]` and `[8, 6, 4, 2, 0]` respectively.
- Indices seen by each worker is then shuffled deterministically. So, the indices seen by each worker may look like: `[5, 7, 3, 1, 9]` and `[4, 6, 2, 0, 8]` respectively.

In addition, the data sampler also provides a mechanism to control the degree of load balancing using the `random_level` argument. This argument modifies the sample complexities randomly so that the sorted dataset is not sorted purely in ascending order of complexities.

The data sampler also allows users to change the sorting and shuffling strategy slightly by working with groups within the dataset. This helps with decoupling any correlation that the sorting of the dataset might introduce with training loss.
- If the user provides the `group_size` argument, the dataset is grouped into `group_size` sized groups.
- The sorting of data samples based on complexities happens within each group as opposed to across the entire dataset.
- The shuffling happens on the group order and not on the dataset sample indices.

Here is an example of data sampler working with groups:
- Assume the data sampler is working with dataset with complexities: `[9, 8, 7, 6, 5, 4, 3, 2, 1, 0]` and the number of workers is 2 and group size is 4.
- The sorted complexities will be: `[6, 7, 8, 9 | 2, 3, 4, 5 | 0, 1]` and sorted dataset indices (sorted on complexities) will be `[3, 2, 1, 0 | 7, 6, 5, 4 | 9, 8]`.
- The group order is then shuffled deterministically so that the indices could be: `[ 7, 6, 5, 4 | 9, 8 | 3, 2, 1, 0]`.
- The data is then distributed among the 2 workers. So each worker will see the data with indices `[7, 5, 9, 3, 1]` and `[6, 4, 8, 2, 0]` respectively.

### CODE REVIEW

File: orttraining/orttraining/python/training/utils/data/sampler.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+            processes in the distributed group. Default: 0.
+        drop_last (bool, optional): if ``True``, then the sampler will drop the
+            tail of the data to make it evenly divisible across the number of
+            shards. If ``False``, the sampler will add extra indices to make
+            the data evenly divisible across the shards. Default: ``False``.
+        random_level (float, optional): A float varies from 0 and 1 that controls the extent
+            of load balance. 0 means the best load balance, while 1 means the opposite.
+    .. warning::
+        In distributed mode, calling the :meth:`set_epoch` method at
+        the beginning of each epoch **before** creating the `DataLoader <https://pytorch.org/docs/stable/data.html?highlight=dataloader#torch.utils.data.DataLoader>`_ iterator
```
Comment (xadupre) [inline orttraining/orttraining/python/training/utils/data/sampler.py:55]: too long line

File: orttraining/orttraining/python/training/utils/data/sampler.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+
+# Implementation is heavily derived from bagua/load_balancing_data_loader.py
+# https://github.com/BaguaSys/bagua/blob/01874a7c3f90904c37c5612a9db866b5d4b8b5ed/bagua/torch_api/contrib/load_balancing_data_loader.py#L12
+class LoadBalancingDistributedSampler:
+    r"""Sampler that balances the data load across workers based on the sample's complexity.
+    This sampler uses a :attr:`complexity_fn` to calculate each sample's computational
+    complexity and make each batch get similar computational complexity.
+    This is useful in scenarios like speech and NLP, where each batch has variable
+    length and distributed training suffers from straggler problem.
+    The usage is similar to `torch.utils.data.DistributedSampler <https://pytorch.org/docs/stable/data.html?highlight=distributedsampler#torch.utils.data.distributed.DistributedSampler>`_,
```
Comment (xadupre) [inline orttraining/orttraining/python/training/utils/data/sampler.py:22]: tool long line

File: orttraining/orttraining/python/training/utils/data/sampler.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+from collections import OrderedDict
+
+
+# Implementation is heavily derived from bagua/load_balancing_data_loader.py
+# https://github.com/BaguaSys/bagua/blob/01874a7c3f90904c37c5612a9db866b5d4b8b5ed/bagua/torch_api/contrib/load_balancing_data_loader.py#L12
+class LoadBalancingDistributedSampler:
+    r"""Sampler that balances the data load across workers based on the sample's complexity.
+    This sampler uses a :attr:`complexity_fn` to calculate each sample's computational
+    complexity and make each batch get similar computational complexity.
+    This is useful in scenarios like speech and NLP, where each batch has variable
```
Comment (xadupre) [inline orttraining/orttraining/python/training/utils/data/sampler.py:20]: an example of complexity function for NLP or speech would help (sequence length?)

File: orttraining/orttraining/python/training/utils/data/sampler.py [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+            # using this Sampler.
+            self.num_samples = dataset_len // self.world_size
+        else:
+            self.num_samples = math.ceil(dataset_len / self.world_size)
+        self.total_size = self.num_samples * self.world_size
+        self.shuffle = shuffle
+        self.seed = seed
+
+        self.sample_complexity_list = [None]*dataset_len
+        for sample_index in range(dataset_len):
```
Comment (xadupre) [inline orttraining/orttraining/python/training/utils/data/sampler.py:125]: Why not moving that part in __iter__ or _sort_shard_and_shuffle_dataset? That would make the constructor a lot faster and the user would not pay a very long cost while building his dataset and his model.

File: orttraining/orttraining/python/training/utils/data/sampler.py [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+            # This is to ensure each rank receives the same amount of data when
+            # using this Sampler.
+            self.num_samples = dataset_len // self.world_size
+        else:
+            self.num_samples = math.ceil(dataset_len / self.world_size)
+        self.total_size = self.num_samples * self.world_size
+        self.shuffle = shuffle
+        self.seed = seed
+
+        self.sample_complexity_list = [None]*dataset_len
```
Comment (xadupre) [inline orttraining/orttraining/python/training/utils/data/sampler.py:124]: You could probably save some space by using numpy array.

File: orttraining/orttraining/python/training/utils/data/sampler.py [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+            will be shuffled. Default: ```None```
+        seed (int, optional): random seed used to shuffle the sampler if
+            :attr:`shuffle=True`. This number should be identical across all
+            processes in the distributed group. Default: 0.
+        drop_last (bool, optional): if ``True``, then the sampler will drop the
+            tail of the data to make it evenly divisible across the number of
+            shards. If ``False``, the sampler will add extra indices to make
+            the data evenly divisible across the shards. Default: ``False``.
+        random_level (float, optional): A float varies from 0 and 1 that controls the extent
+            of load balance. 0 means the best load balance, while 1 means the opposite.
```
Comment (xadupre) [inline orttraining/orttraining/python/training/utils/data/sampler.py:52]: opposite or just more random? From what I understood, it would be random if random_level is high (>> 1).

**Comment by ytaous** [comment]
A couple questions:
1. have anyone tried the util on any model,if so, what's the perf % we are expecting?
2. what's the trade-off, i.e., memory penalty and wait time, when using the util?

**Review by ytaous — DISMISSED** [review]

**Comment by baijumeswani** [comment]
> A couple questions:
> 
> 1. has anyone tried the util on any model,if so, what's the perf % we are expecting?

I have tried it on one of our internal models and observed around 3~4% improvement with PyTorch. I am in the process of evaluating this on another model right now.

> 2. what's the trade-off, i.e., memory penalty and wait time, when using the util?

I don't think there is any penalty on the memory. The data sampler holds onto a list of pairs whose values are the index of the sample in the dataset and the complexity of this sample. From my observations, this sampler does not incur any significant memory penalty.
The sampler also sorts this list before every epoch. So, it might incur some wait time penalty. But on the model I ran, I was able to observe around 3~4% performance improvement despite the sorting.

**Review by ytaous — APPROVED** [review]

**Review by SherlockNoMad — APPROVED** [review]
